# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [5]:
def addTwo(x):
    return x +2
print(addTwo(4))

# The lambda version...
# Without using a variable to pass in data
print((lambda x: x + 2)(4))

# Using a variable
a_num = 4
print((lambda x: x + 2)(a_num))

6
6
6


#### Saving to a Variable

In [7]:
f_test = lambda x: x + 2

print(f_test(4))

6


#### Multiple Inputs

In [13]:
# multiple inputs with no variable

print((lambda x, y, z: x * y * z)(4, 6, 7))

# Let's make it a function! (passing lambda into a function)

a_test = lambda x, y, z: x * y * z
print(a_test(4, 5, 6))

168
120


#### Passing a Lambda into a Function

In [8]:
def multiply(f, num):
    """
    'f' will expect a lambda function,
    'num' will expect a number
    """
    return f(num)

multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [18]:
def multiply_tst(num):
    return num * 2

# Function within a function:
def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(f_return(4))

# Writing lambda function returned from a regular function (saves lines?)
def returnLamb(b,c):
    return lambda x, a: x + a + b + c
r_lamb = returnLamb(4,5)
print(r_lamb(5,5))

8
19


#### If Statements within Lambdas

In [21]:
# Kind of like list comprehension.
# Lambda = True if (condition) is met, otherwise False

f_condition = lambda num: num * 2 if num > 10 else num + 4
print(f_condition(3))
print(f_condition(10))
print(f_condition(30))
print(f_condition(25))

7
14
60
50


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [22]:
# Write the lambda (anonymous) function:
f_cubed = lambda x: x * x * x

# Assign it to the variable f:
f = f_cubed

# test:
print(f(3))

27


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [24]:
def squared(num, num2):
    if num > 10 or num2 < 10:
        return num **2, num2 **2
    else:
        return num, num2

number = [3, 5, 10, 29, 30, 4]
more_numbers = [30, 27, 4, 10, 15]

squared_nums_map = list(map(squared, number, more_numbers))
print(squared_nums_map)

[(3, 30), (5, 27), (100, 16), (841, 100), (900, 225)]


#### Using Lambda's with Map

In [25]:
# map(lambda x: do something, iterable)
# Using lambda in map happens inline (or on one line), less lines = better/cooler
# list(map(lambda x: x ** 2, nums)) is our basic formula

squared_nums_lamb = list(map(lambda x,y: (x ** 2, y ** 2) if x > 10 or y < 10 else (x,y), number, more_numbers))
print(squared_nums_lamb)

[(3, 30), (5, 27), (100, 16), (841, 100), (900, 225)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [28]:
# Step one, double each number
# Step two, subtract one

doubled_nums_lamb = list(map(lambda x,y: ((x * 2) - 1, (y * 2) - 1), number, more_numbers))
print(number)
print(more_numbers)
print(doubled_nums_lamb)
print("I like this so much more than Regex")

[3, 5, 10, 29, 30, 4]
[30, 27, 4, 10, 15]
[(5, 59), (9, 53), (19, 7), (57, 19), (59, 29)]
I like this so much more than Regex


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [37]:
names = ['Bob', 'Andy', 'Max', 'Evan', 'Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    elif name[0].lower() == 'e':
        return True
    else:
        return False

new_names_list = list(filter(a_names, names))
print(new_names_list)

['Andy', 'Evan', 'Angelica']


#### Using Lambda's with Filter()

In [43]:
new_names_lambda = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
# Pay attention to the way the "True" and "if" switch around in syntax

print(new_names_lambda)
                        
                        

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [54]:
from statistics import mean

# Then, I have to make a list of random numbers
my_numbers = [2, 3, 4, 5, 6, 7, 8, 9]

print(mean(my_numbers))

# Now, I'll try out a regular function
def a_mean(num):
        if num > 5.5:
            return True
        else:
            return False

new_numbers_list = list(filter(a_mean, my_numbers))
print(new_numbers_list)

# That works! Let's try lambda.
new_numbers_lambda = list(filter(lambda x: True if x > 5.5 else False, my_numbers))
print(new_numbers_lambda)
print("Mission CLEAR")

5.5
[6, 7, 8, 9]
[6, 7, 8, 9]
Mission CLEAR


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [57]:
from functools import reduce
# reduce(function, iterable)
# This isn't really used anymore, but you may see in older code.
# Basically applies function to entire list, I think?

list_1 = [2, 4, 6, 8, 10]

def addNums(num1, num2):
    return num1 + num2

result_add = reduce(addNums, list_1)
print(result_add)

# Subtraction looks like:

def subtractNums(num1, num2):
    return num1 - num2

result_sub = reduce(subtractNums, list_1)
print(result_sub)

30
-26


#### Using Lambda's with Reduce()

In [59]:
result_lamb = reduce(lambda x, y: x + y, list_1)
print(result_lamb)

sub_lamb = reduce(lambda x, y: x - y, list_1)
print(sub_lamb)

30
-26


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [60]:
my_list = [1, 2, 3, 4]

mult_lamb = reduce(lambda x, y: x * y, my_list)
print(mult_lamb)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [66]:
# 1.
# def addNums(num):
# #     This is the base case. It's SUPER IMPORTANT.
#     if num <= 1:
#         print("addNums(1) = 1")
#         return num
#     else:
#         print(f"addNums({num}) = addNums({num - 1})")
# addNums(2)

# It prints that way because I haven't used recursion yet.


# 2.
def addNums(num):
#     This is the base case. It's SUPER IMPORTANT.
    if num <= 1:
        print("addNums(1) = 1")
        return num
    else:
        print(f"addNums({num}) = addNums({num - 1})")
#         Here's the recursion part:
        return num + addNums(num - 1)
addNums(5)

addNums(5) = addNums(4)
addNums(4) = addNums(3)
addNums(3) = addNums(2)
addNums(2) = addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [67]:
# factorial 5 = 5*4*3*2*1

def factorial(num):
    if num <= 1:
        return 1
    else:
        return num + factorial(num-1)

factorial(9)

45

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [76]:
# "argument given" = the (9) in the factorial(9) above. Don't forget again!

def subtract(num):
    if num <= 1:
        return 1
    else:
        return num + subtract(num - 1)

subtract(100)

5050

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [83]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start
#         That right there denotes a generator, that yield command. It's like return, but it doesn't stop the function. 
        start += step

for i in my_range(20, start = 2):
    my_generator_value = i
    print(my_generator_value)
    
# Here's a random conversion to range I did:
for i in range(1, 20, 2):
    my_generator_value = i
    print(my_generator_value)
    

2
4
6
8
10
12
14
16
18


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [77]:
# -----------------------NOTES---------------------------------
# def my_range(stop, start, step = 2):
#     while start < stop:
#         yield start
# #         That right there denotes a generator, that yield command. It's like return, but it doesn't stop the function. 
#         start += step

# for i in my_range(20, start = 2):
#     my_generator_value = i
#     print(my_generator_value)
    
# Here's a random conversion to range I did:
# for i in range(1, 20, 2):
#     my_generator_value = i
#     print(my_generator_value)
    

def liz_range(stop, start, step):
    while start > stop:
        yield start
        start -= step

for i in liz_range(0, 10, 1):
    liz_generator_value = i*i
    print(liz_generator_value)
print("Thanks Donovan!")

100
81
64
49
36
25
16
9
4
1
Thanks Donovan!


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [78]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

# strip method - strip the list!

new_places_lambda = list(filter(lambda place: place.strip(' '), places))



print(new_places_lambda)
print("Thanks Shay!")


['Argentina', 'San Diego', 'Boston', 'New York']
Thanks Shay!


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [31]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

# Sort command first, then insert lambda function that splits the name in half by space.Sort by key

# type(author) 

author.sort(key = lambda name: name.split(" ")[-1].lower())

print(author)



['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [43]:

# ------------------------NOTES----------------------------------------------------
# map(lambda x: do something, iterable)
# Using lambda in map happens inline (or on one line), less lines = better/cooler
# list(map(lambda x: x ** 2, nums)) is an example of our basic formula

# number = [3, 5, 10, 29, 30, 4]
# more_numbers = [30, 27, 4, 10, 15]
# squared_nums_lamb = list(map(lambda x,y: (x ** 2, y ** 2) if x > 10 or y < 10 else (x,y), number, more_numbers))
# print(squared_nums_lamb)

# ----------------------SECOND EXAMPLES-------------------------------------------
# Step one, double each number
# Step two, subtract one

# doubled_nums_lamb = list(map(lambda x,y: ((x * 2) - 1, (y * 2) - 1), number, more_numbers))
# print(number)
# print(more_numbers)
# print(doubled_nums_lamb)
# print("I like this so much more than Regex")

# --------------------PROBLEM----------------------------------------------------

# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angeles",44),("Miami",29)]

places_celcius = (list(map(lambda x: (x[0], (9/5) * x[1] + 32), places)))

print("This is the temperature in Celicus: ", (places))
print("This is the temperature in Farenheit: ", places_celcius)



This is the temperature in Celicus:  [('Nashua', 32), ('Boston', 12), ('Los Angeles', 44), ('Miami', 29)]
This is the temperature in Farenheit:  [('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [54]:
# ----------------NOTES-----------------------------------
# def addNums(num):
# #     This is the base case. It's SUPER IMPORTANT.
#     if num <= 1:
#         print("addNums(1) = 1")
#         return num
#     else:
#         print(f"addNums({num}) = addNums({num - 1})")
# #         Here's the recursion part:
#         return num + addNums(num - 1)
# addNums(5)

def Iteration(num):
#     base case:
    if num == 0:
        return 0
    elif num == 1:
        return 1
    else:
        return Iteration(num - 1) + Iteration(num - 2)

def fib(n):
    for i in range(n + 1):
        print("Iteration({i}): {x}".format(i = i, x = Iteration(i + 1)))
fib(5)

Iteration(0): 1
Iteration(1): 1
Iteration(2): 2
Iteration(3): 3
Iteration(4): 5
Iteration(5): 8
